src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1577,50 +1556,63 @@ where
 /// TODO: This function is a bit of a mess rn but hopefully this function makes
 /// the existing mess more obvious and points towards ways to improve it.
 fn handle_message<K, V, D>(
-    message: SourceMessage<K, V, D>,
+    message: Result<SourceMessage<K, V>, SourceReaderError>,
+    time: (PartitionId, MzOffset),
+    diff: D,
     bytes_read: &mut usize,
     cap_set: &CapabilitySet<Timestamp>,
-    output: &mut OutputHandle<
+    output_handle: &mut OutputHandle<
         Timestamp,
         (usize, Result<SourceOutput<K, V, D>, SourceError>),
         Tee<Timestamp, (usize, Result<SourceOutput<K, V, D>, SourceError>)>,
     >,
     metric_updates: &mut HashMap<PartitionId, (MzOffset, Timestamp, i64)>,
     ts: Timestamp,
+    source_id: GlobalId,
 ) where
     K: timely::Data + MaybeLength,
     V: timely::Data + MaybeLength,
     D: timely::Data,
 {
-    let partition = message.partition.clone();
-    let offset = message.offset;
-
-    // Note: empty and null payload/keys are currently treated as the same
-    // thing.
-    let key = message.key;
-    let out = message.value;
-    // Entry for partition_metadata is guaranteed to exist as messages are only
-    // processed after we have updated the partition_metadata for a partition
-    // and created a partition queue for it.
-    if let Some(len) = key.len() {
-        *bytes_read += len;
-    }
-    if let Some(len) = out.len() {
-        *bytes_read += len;
-    }
+    let partition = time.0.clone();
+    let offset = time.1;
+
+    let output = match message {
+        Ok(message) => {
+            // Note: empty and null payload/keys are currently treated as the same thing.
+            if let Some(len) = message.key.len() {
+                *bytes_read += len;
+            }
+            if let Some(len) = message.value.len() {
+                *bytes_read += len;
+            }
+
+            (
+                message.output,
+                Ok(SourceOutput::new(
+                    message.key,
+                    message.value,
+                    offset,
+                    message.upstream_time_millis,
+                    time.0,
+                    message.headers,
+                    diff,
+                )),
+            )
+        }
+        Err(err) => {
+            let err = SourceError {
+                source_id,
+                error: err.inner,
+            };
+            // XXX(petrosagg): errors should be attributed to a specific output by the source impl
+            // instead of hardcoding it to output zero
```

Comment:
Is this a subsources thing? (Is this behaviour any different than what happens today?)